### PR TITLE
Add a JWT-SVID authentication method

### PR DESCRIPTION
**Pull request by hug-dev**
Using this method, the client will fetch its JWT-SVID and send it in the
authentication field.
The socket endpoint is found through the SPIFFE_ENDPOINT_SOCKET
environment variable.

### CODE REVIEW

**Comment by hug-dev** [comment]
This won't work until the changes are pushed in some `rust-spiffe` repository 🥇

File: src/error.rs [hunk 1]
```diff
@@ -42,6 +42,8 @@ pub enum ClientErrorKind {
     NoAuthenticator,
     /// Required parameter was not provided
     MissingParam,
+    /// Error while using the SPIFFE Workload API
+    Spiffe,
```
Comment (ionut-arm) [inline src/error.rs:46]: If the Spiffe error object implements `std::error::Error` you could/should add it here like we do with `Interface` above. If it doesn't.... you should probs put that on the TODO list 🤡

File: src/auth.rs [hunk 2]
```diff
@@ -28,18 +37,40 @@ impl Authentication {
...
+                            e
+                        );
+                        Error::Client(ClientErrorKind::Spiffe)
+                    })?,
+                    None,
+                );
+                let audience = String::from("parsec");
+
+                let result = client.fetch(audience, None).map_err(|e| {
+                    error!("Error while fetching the JWT-SVID ({}).", e);
```
Comment (ionut-arm) [inline src/auth.rs:70]: Can that error object contain any sensitive data?

File: src/auth.rs [hunk 3]
```diff
@@ -28,18 +37,40 @@ impl Authentication {
...
             Authentication::UnixPeerCredentials => {
                 let current_uid = users::get_current_uid();
-                RequestAuth::new(current_uid.to_le_bytes().to_vec())
+                Ok(RequestAuth::new(current_uid.to_le_bytes().to_vec()))
+            }
+            Authentication::JwtSvid => {
+                let client = JWTClient::new(
+                    &env::var("SPIFFE_ENDPOINT_SOCKET").map_err(|e| {
+                        error!(
+                            "Can not read the SPIFFE_ENDPOINT_SOCKET environment variable ({}).",
```
Comment (ionut-arm) [inline src/auth.rs:60]:
```suggestion
                            "Cannot read the SPIFFE_ENDPOINT_SOCKET environment variable ({}).",
```
🤓

**Review by ionut-arm — COMMENTED** [review]
Thanks, it looks good overall, just a few minor comments

File: src/error.rs (hunk 1 above)
Comment (hug-dev) [reply src/error.rs:46]: Ah yes good point... There is an error, I will put it there.

File: src/auth.rs (hunk 2 above)
Comment (hug-dev) [reply src/auth.rs:70]: I don't think so, it's mostly related with the transport

File: src/auth.rs (hunk 3 above)
Comment (hug-dev) [reply src/auth.rs:60]: 🤓

**Comment by hug-dev** [comment]
Addressed comment in a new commit for easier review.

File: src/core/testing/core_tests.rs [hunk 4]
```diff
@@ -179,7 +179,10 @@ fn core_provider_for_crypto_test() {
         .psa_destroy_key(String::from("random key"))
         .expect_err("Expected a failure!!");
 
-    assert_eq!(res, Error::Client(ClientErrorKind::InvalidProvider));
+    assert!(matches!(
+        res,
+        Error::Client(ClientErrorKind::InvalidProvider)
+    ));
```
Comment (ionut-arm) [inline src/core/testing/core_tests.rs:185]: 💯

**Review by ionut-arm — APPROVED** [review]
👍

**Comment by hug-dev** [comment]
Used this [branch](https://github.com/hug-dev/rust-spiffe/pull/1) to make it compile now. Added a feature as was agreed in the [community meeting](https://github.com/parallaxsecond/community/blob/master/minutes/2020-10-13.md#minutes).

File: src/auth.rs [hunk 5]
```diff
@@ -28,18 +35,48 @@ impl Authentication {
...
+                    None,
+                    None,
+                );
+                let audience = String::from("parsec");
+
+                let result = client.fetch(audience).map_err(|e| {
+                    error!("Error while fetching the JWT-SVID ({}).", e);
+                    Error::Client(ClientErrorKind::Spiffe(e))
+                })?;
+                Ok(RequestAuth::new(result.svid().as_bytes().into()))
```
Comment (ionut-arm) [inline src/auth.rs:79]: Do you need to zeroize the contents of `result`, given that they contain a token?

File: tests/ci.sh [hunk 6]
```diff
@@ -13,6 +13,7 @@ set -euf -o pipefail
 ################
 RUST_BACKTRACE=1 cargo build
 RUST_BACKTRACE=1 cargo build --features testing
+RUST_BACKTRACE=1 cargo build --no-default-features
```
Comment (ionut-arm) [inline tests/ci.sh:16]: 👍

**Review by ionut-arm — COMMENTED** [review]
🚀

File: src/auth.rs (hunk 5 above)
Comment (hug-dev) [reply src/auth.rs:79]: I will implement `Zeroize` on drop on the `Jwt` structure to make sure it is zeroized all the time! Will do that directly in the `rust-spiffe` structure.

Comment (hug-dev) [reply src/auth.rs:79]: Did it [here](https://github.com/hug-dev/rust-spiffe/blob/7a6fd03f81fb194feaf5fe6c19e07db8843c6b72/spiffe/src/svid/jwt.rs#L21)!

Comment (ionut-arm) [reply src/auth.rs:79]: Cheers 💯

**Review by ionut-arm — APPROVED** [review]

**Pull request merged**